lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -2936,7 +2941,78 @@ ParseResult FIRStmtParser::parsePrintf() {
 
   locationProcessor.setLoc(startTok.getLoc());
 
-  auto formatStrUnescaped = FIRToken::getStringValue(formatString);
+  // Validate the format string and remove any "special" substitutions.  Only do
+  // this for FIRRTL versions >= 4.2.0.  If at a different FIRRTL version, then
+  // just parse this as if it was a string.
+  SmallVector<Attribute, 4> specialSubstitutions;
+  SmallString<64> validatedFormatString;
+  SmallVector<Value, 4> operands;
+  if (version < missingSpecFIRVersion) {
+    validatedFormatString = formatString;
+    operands.append(specOperands);
+  } else {
+    for (size_t i = 0, e = formatString.size(), opIdx = 0; i != e; ++i) {
+      auto c = formatString[i];
+      switch (c) {
+      // FIRRTL percent format strings.  If this is actually a format string,
+      // then grab one of the "spec" operands.
+      case '%': {
+        validatedFormatString.push_back(c);
+        c = formatString[++i];
+        switch (c) {
+        case 'b':
+        case 'd':
+        case 'x':
+          operands.push_back(specOperands[opIdx++]);
+          break;
+        // Let anything we don't know about through.  This allows for wildcat
+        // use of `%m` if a user wants.
+        default:
+          break;
+        }
+        validatedFormatString.push_back(c);
+        break;
+      }
```

Comment:
This is really cool! We could spit out dedicated `firrtl.fstring.int ...` ops for these, and then concatenate everything into a single `!firrtl.fstring` that we pass to `firrtl.printf`.

Reply:
This is _exactly_ the direction I would like to move this in in the future. Thanks for the points to `sim` dialect and Rust's `format_args` that precipitated this push.